theon/src/main.rs

```diff
@@ -13,21 +13,175 @@
+}
+
+/// Main entry point for the loader.
+///
+/// When we enter `main()`, the first 4GiB of the
+/// physical address space are mapped R/W at
+/// arch::HYPER_BASE_VADDR.  Note that the memory
+/// regions that make up both the binary archive
+/// as well as our load regions are mapped within
+/// this region, so we can address them via pointers.
 #[cfg_attr(not(test), start, no_mangle)]
 pub extern "C" fn main(mbinfo_phys: u64) -> ! {
-    x86_64::init::start(mbinfo_phys);
+    let multiboot = x86_64::init::start(mbinfo_phys);
+    let crate::x86_64::multiboot1::InitInfo { memory_regions, regions, modules } = multiboot.info();
+    core::mem::drop(memory_regions);
+    uart::panic_println!("end = {:016x}", end_addr());
+    uart::panic_println!("regions: {:#x?}", regions);
+    let bins = modules.iter().find(|m| m.name == Some("bin.a")).expect("found 'bin.a' in modules");
+    let archive = goblin::archive::Archive::parse(bins.bytes).expect("cannot parse bin.a");
+    uart::panic_println!("Binary archive: {:#x?}", archive);
+    clear_binary_load_region();
+    for &(name, addr, typ) in BINARY_TABLE {
+        let bytes = archive.extract(name, bins.bytes).expect("cannot extract elf");
+        let region_end = addr.offset(BINARY_IMAGE_MEMORY_SIZE);
+        load(name, typ, bytes, addr..region_end).expect("loaded binary");
+    }
     unsafe { core::arch::asm!("int3") };
     panic!("main: trapstubs = {:#x}", arch::trap::stubs as usize);
 }
 
+/// Zeroes the memory region that binaries are loaded into.
+fn clear_binary_load_region() {
+    let start = BINARY_LOAD_REGION_START.vaddress();
+    let end = BINARY_LOAD_REGION_END.vaddress();
+    unsafe { core::ptr::write_bytes(start as *mut u8, 0, end - start) };
+}
+
+/// Loads the named binary of the given type into given physical region.
+fn load(name: &str, typ: BinaryType, bytes: &[u8], region: Range<HPA>) -> Result<PF4K> {
+    use arch::{Page, Page4K};
+    let elf = goblin::elf::Elf::parse(bytes).expect("cannot parse elf");
+    uart::panic_println!(
+        "ELF for {:#?} ({:?}@{:x?}): {:#x?}",
+        name,
+        typ,
+        region,
+        elf.program_headers
+    );
+    let mut regions = alloc::vec::Vec::new();
+    let mut headers = alloc::vec::Vec::new();
+    for header in
+        elf.program_headers.iter().filter(|h| h.p_type == goblin::elf::program_header::PT_LOAD)
+    {
+        let vm = header.vm_range();
+        // All Hypatia binaries require that loadable sections
+        // are aligned on 4KiB boundaries.
+        assert_eq!(vm.start % 4096, 0);
+        assert!(vm.start < vm.end);
+        regions.push(V4KA::new(vm.start)..V4KA::new_round_up(vm.end));
+        headers.push(header);
+    }
+    let base = region.start.vaddress() as *mut u8;
+    let len = region.end.vaddress() - region.start.vaddress();
+    let heap = unsafe { core::slice::from_raw_parts_mut(base, len) };
+    let bump = allocator::BumpAlloc::new(heap);
+    let allocator = || {
```

Comment:
Rename to "allocate" so later calls look like "let mem = allocate().expect();"

Reply:
Done.

(Much nicer.)